stdlib/Statistics/src/Statistics.jl

```diff
@@ -73,7 +73,29 @@ function mean(f, itr)
     end
     return total/count
 end
-mean(f, A::AbstractArray) = sum(f, A) / length(A)
+
+"""
+    mean(f::Function, A::AbstractArray, [dims])
+
+Apply the function `f` to each element of `A` across `dims` and take the mean.
+
+```jldoctest
+julia> mean(√, [1, 2, 3])
+1.3820881233139908
+
+julia> mean([√1, √2, √3])
+1.3820881233139908
+
+julia> mean(√, [1 2 3; 4 5 6], dims = 2)
+2×1 Array{Float64,2}:
+ 1.3820881233139908
+ 2.2285192400943226
+```
+"""
+mean(f, A::AbstractArray; dims = :) = _mean(f, A, dims)
+
+_mean(f, A::AbstractArray, ::Colon) = sum(f, A) / length(A)
+_mean(f, A::AbstractArray, dims) = sum(f, A, dims = dims) / prod(size(A, i) for i in dims)
```

Comment:
Would be good to check `@inferred` for this tuple case.

Reply:
Also, what happens if dims is an empty tuple?

Reply:
You will get `ERROR: ArgumentError: reducing over an empty collection is not allowed` if you pass `dims=()` (which is supported by `sum`).  Instead of the `prod` call maybe you can use:
```jl
mapreduce(i -> size(A, i), *, dims; init=1)
```